components/motor/gpio/loop.go

```diff
@@ -60,27 +75,259 @@ func (m *EncodedMotor) setupControlLoop() error {
 		options.NeedsAutoTuning = true
 	}
 
-	pl, err := control.SetupPIDControlConfig(convertedControlParams, m.Name().ShortName(), options, m, m.logger)
+	pl, err := control.SetupPIDControlConfig(convertedControlParams, cm.Name().ShortName(), options, cm, cm.logger)
 	if err != nil {
 		return err
 	}
 
-	m.controlLoopConfig = pl.ControlConf
-	m.loop = pl.ControlLoop
-	m.blockNames = pl.BlockNames
+	cm.controlLoopConfig = pl.ControlConf
+	cm.loop = pl.ControlLoop
+	cm.blockNames = pl.BlockNames
 
 	return nil
 }
 
-func (m *EncodedMotor) startControlLoop() error {
-	loop, err := control.NewLoop(m.logger, m.controlLoopConfig, m)
+func (cm *controlledMotor) startControlLoop() error {
+	loop, err := control.NewLoop(cm.logger, cm.controlLoopConfig, cm)
 	if err != nil {
 		return err
 	}
 	if err := loop.Start(); err != nil {
 		return err
 	}
-	m.loop = loop
+	cm.loop = loop
+
+	return nil
+}
+
+func setupMotorWithControls(
+	_ context.Context,
+	m *Motor,
+	enc encoder.Encoder,
+	cfg resource.Config,
+	logger logging.Logger,
+) (motor.Motor, error) {
+	conf, err := resource.NativeConfig[*Config](cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	tpr := float64(conf.TicksPerRotation)
+	if tpr == 0 {
+		tpr = 1.0
+	}
+
+	cm := &controlledMotor{
+		Named:            cfg.ResourceName().AsNamed(),
+		logger:           logger,
+		opMgr:            operation.NewSingleOperationManager(),
+		ticksPerRotation: tpr,
+		real:             m,
+		enc:              enc,
+	}
+
+	// setup control loop
+	if conf.ControlParameters != nil {
+		if err := cm.setupControlLoop(conf); err != nil {
+			return nil, err
+		}
+	}
+
+	return cm, nil
+}
+
+type controlledMotor struct {
+	resource.Named
+	resource.AlwaysRebuild
+	logger                  logging.Logger
+	opMgr                   *operation.SingleOperationManager
+	activeBackgroundWorkers sync.WaitGroup
+
+	offsetInTicks    float64
+	ticksPerRotation float64
+
+	mu   sync.RWMutex
+	real *Motor
+	enc  encoder.Encoder
+
+	controlLoopConfig control.Config
+	blockNames        map[string][]string
+	loop              *control.Loop
+}
+
+// SetPower sets the percentage of power the motor should employ between -1 and 1.
+// Negative power implies a backward directional rotational.
+func (cm *controlledMotor) SetPower(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
+	cm.opMgr.CancelRunning(ctx)
+	if cm.loop != nil {
+		cm.loop.Pause()
+	}
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
```

Comment:
do we need this lock? we aren't changing any state in `cm` and `real.SetPower` already has a lock